SofaKernel/framework/sofa/defaulttype/Mat.h

```diff
@@ -324,12 +324,30 @@ class Mat : public helper::fixed_array<VecNoInit<C,real>,L>
         return id;
     }
 
+
+    template<int S>
+    static bool canSelfTranspose(const Mat<S, S, real>& lhs, const Mat<S, S, real>& rhs)
+    {
+        return &lhs == &rhs;
+    }
+
+    template<int I, int J>
+    static bool canSelfTranspose(const Mat<I, J, real>& /*lhs*/, const Mat<J, I, real>& /*rhs*/)
+    {
+        return false;
+    }
+
     /// Set matrix as the transpose of m.
     void transpose(const Mat<C,L,real> &m)
     {
-        for (int i=0; i<L; i++)
-            for (int j=0; j<C; j++)
-                this->elems[i][j]=m[j][i];
+        if (canSelfTranspose(*this, m))
+            *this = transposed();
```

Comment:
rather than creating a full temporary, and then copying everyone, what you could do is to iterate only on a triangular matrix and to swap with the other triangular. Something like:

`for (int i=0; i<L; i++)
 for (int j=i+1; j<C; j++)
     std::swap( this->elems[i][j], this->elems[j][i] );
`

Reply:
@matthieu-nesme : copied the behavior of `void transpose()`.
Thanks for your review!